build.assets/tooling/lib/github/workflows.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package github
+
+import (
+	"context"
+	"log"
+	"time"
+
+	"github.com/google/go-github/v41/github"
+	"github.com/gravitational/trace"
+)
+
+func setsAreEqual(a, b map[int64]struct{}) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for k := range a {
+		if _, present := b[k]; !present {
+			return false
+		}
+	}
+	return true
+}
+
+func firstUniqueInSet(newSet, baseSet map[int64]struct{}) (int64, bool) {
+	for k, _ := range newSet {
+		if _, present := baseSet[k]; !present {
+			return k, true
+		}
+	}
+	return 0, false
+}
+
+// ListWorkflowRuns returns a set of RunIDs, representing the set of all for
+// workflow runs created since the supplied start time.
+func (gh *ghClient) ListWorkflowRuns(ctx context.Context, owner, repo, path, ref string, since time.Time) (map[int64]struct{}, error) {
+	listOptions := github.ListWorkflowRunsOptions{
+		ListOptions: github.ListOptions{
+			PerPage: 100,
+		},
+		Branch:  ref,
+		Created: ">" + since.Format(time.RFC3339),
+	}
+
+	runIDs := make(map[int64]struct{})
+
+	for {
+		runs, resp, err := gh.client.Actions.ListWorkflowRunsByFileName(ctx, owner, repo, path, &listOptions)
+		if err != nil {
+			return nil, trace.Wrap(err, "Failed to fetch runs")
+		}
+
+		for _, r := range runs.WorkflowRuns {
+			runIDs[r.GetID()] = struct{}{}
+		}
+
+		if resp.NextPage == 0 {
+			break
+		}
+
+		listOptions.Page = resp.NextPage
+	}
+
+	return runIDs, nil
+}
+
+// TriggerDispatchEvent triggers a workflow_dispatch event in the target
+// repository and waits for a workflow to be started in response. Note that
+// this method requires that the GitHub and client clocks are roughly in sync.
+func (gh *ghClient) TriggerDispatchEvent(ctx context.Context, owner, repo, workflow, ref string, inputs map[string]interface{}) (*github.WorkflowRun, error) {
+	// There is no way that I know of to 100% accurately detect which workflow runs
+	// are created in response to a workflow_dispatch event. We can make a vey good
+	// guess, though, by looking at what workflow runs (with matching filename and
+	// source references) start immediately after we issue the event - so that's
+	// what we do here.
```

Comment:
[this](https://stackoverflow.com/a/69500478/5859521) SO answer explains in idea of putting an ID as an input and finding the run with that ID, would it be feasible for us?